File: tests/drift/integration.test.ts
/**
 * End-to-end gate scenario tests (epic fitness function): S3, S4, S6 stay
 * green; persist split-brain (S15-extension) stays green across SIGINT
 * injection.
 *
 * These integrate detect → decideGate → applyGate as the swap orchestrator
 * (E5) will compose them. Per-step unit coverage lives in detect.test.ts,
 * gate.test.ts, apply.test.ts.
 */

import { promises as fs } from "node:fs";
import * as path from "node:path";

import { afterEach, beforeEach, describe, expect, it } from "vitest";

import { applyGate } from "../../src/drift/apply.js";
import { detectDrift } from "../../src/drift/detect.js";
import { decideGate } from "../../src/drift/gate.js";
import type { MergedFile } from "../../src/merge/types.js";
import { resolve } from "../../src/resolver/resolve.js";
import { listSnapshots } from "../../src/state/backup.js";
import { materialize } from "../../src/state/materialize.js";
import { buildStatePaths } from "../../src/state/paths.js";
import { readStateFile } from "../../src/state/state-file.js";
import { makeFixture } from "../helpers/fixture.js";

async function setupTwoProfiles() {
  const fx = await makeFixture({
    profiles: {
      base: {
        manifest: { name: "base" },
        files: { "CLAUDE.md": "BASE\n", "agents/a.md": "AGENT-A" },
      },
      leaf: {
        manifest: { name: "leaf", extends: "base" },
        files: { "CLAUDE.md": "LEAF\n" },
      },
      other: {
        manifest: { name: "other" },
        files: { "CLAUDE.md": "OTHER\n" },
      },
    },
  });
  const leafPlan = await resolve("leaf", { projectRoot: fx.projectRoot });
  const leafMerged: MergedFile[] = [
    {
      path: "CLAUDE.md",
      bytes: Buffer.from("BASE\nLEAF\n"),
      contributors: ["base", "leaf"],
      mergePolicy: "concat",
      destination: ".claude",
    },
    {
      path: "agents/a.md",
      bytes: Buffer.from("AGENT-A"),
      contributors: ["base"],
      mergePolicy: "last-wins",
      destination: ".claude",
    },
  ];
  const otherPlan = await resolve("other", { projectRoot: fx.projectRoot });
  const otherMerged: MergedFile[] = [
    {
      path: "CLAUDE.md",
      bytes: Buffer.from("OTHER\n"),
      contributors: ["other"],
      mergePolicy: "concat",
      destination: ".claude",
    },
  ];
  return { fx, leafPlan, leafMerged, otherPlan, otherMerged };
}

describe("E4 integration scenarios", () => {
  let ctx: Awaited<ReturnType<typeof setupTwoProfiles>> | undefined;

  beforeEach(async () => {
    ctx = await setupTwoProfiles();
  });
  afterEach(async () => {
    await ctx?.fx.cleanup();
  });

  it("S3: drift gate — discard → live tree replaced, edits in snapshot", async () => {
    const paths = buildStatePaths(ctx!.fx.projectRoot);
    await materialize(paths, ctx!.leafPlan, ctx!.leafMerged);
    await fs.writeFile(
      path.join(paths.claudeDir, "CLAUDE.md"),
      "DRIFT FROM USER\n",
    );

    const report = await detectDrift(paths);
    expect(report.entries).toHaveLength(1);

    const decision = decideGate({ report, mode: "interactive" });
    // Interactive + drift + no flag → prompt. Simulate the user picking
    // discard.
    expect(decision.kind).toBe("prompt");
    const result = await applyGate("discard", {
      paths,
      plan: ctx!.otherPlan,
      merged: ctx!.otherMerged,
      activeProfileName: report.active,
    });

    expect(result.action).toBe("materialized");
    expect(result.backupSnapshot).not.toBeNull();
    expect(
      await fs.readFile(path.join(paths.claudeDir, "CLAUDE.md"), "utf8"),
    ).toBe("OTHER\n");
    expect(
      await fs.readFile(path.join(result.backupSnapshot!, "CLAUDE.md"), "utf8"),
    ).toBe("DRIFT FROM USER\n");
    const snaps = await listSnapshots(paths);
    expect(snaps).toHaveLength(1);
  });

  it("S4: drift gate — persist → live tree saved into active profile, then swap", async () => {
    const paths = buildStatePaths(ctx!.fx.projectRoot);
    await materialize(paths, ctx!.leafPlan, ctx!.leafMerged);
    // User adds a brand-new file plus edits CLAUDE.md.
    await fs.writeFile(
      path.join(paths.claudeDir, "CLAUDE.md"),
      "EDITED LEAF\n",
    );
    await fs.writeFile(path.join(paths.claudeDir, "scratch.md"), "scratch\n");

    const report = await detectDrift(paths);
    expect(report.entries.map((e) => e.relPath).sort()).toEqual([
      "CLAUDE.md",
      "scratch.md",
    ]);

    const result = await applyGate("persist", {
      paths,
      plan: ctx!.otherPlan,
      merged: ctx!.otherMerged,
      activeProfileName: report.active,
    });
    expect(result.action).toBe("persisted-and-materialized");

    // Active profile dir now reflects live state at moment of persist.
    const persistedDir = path.join(
      ctx!.fx.projectRoot,
      ".claude-profiles",
      "leaf",
      ".claude",
    );
    expect(
      await fs.readFile(path.join(persistedDir, "CLAUDE.md"), "utf8"),
    ).toBe("EDITED LEAF\n");
    expect(
      await fs.readFile(path.join(persistedDir, "scratch.md"), "utf8"),
    ).toBe("scratch\n");
    // Live tree is now the new profile.
    expect(
      await fs.readFile(path.join(paths.claudeDir, "CLAUDE.md"), "utf8"),
    ).toBe("OTHER\n");
    // No snapshot taken on persist path (only discard takes one).
    const snaps = await listSnapshots(paths);
    expect(snaps).toEqual([]);
  });

  it("S6: drift gate — abort → no FS change, state unchanged", async () => {
    const paths = buildStatePaths(ctx!.fx.projectRoot);
    await materialize(paths, ctx!.leafPlan, ctx!.leafMerged);
    await fs.writeFile(
      path.join(paths.claudeDir, "CLAUDE.md"),
      "DRIFT\n",
    );
    const stateBefore = await readStateFile(paths);
    const beforeBytes = await fs.readFile(
      path.join(paths.claudeDir, "CLAUDE.md"),
      "utf8",
    );

    const report = await detectDrift(paths);
    const result = await applyGate("abort", {
      paths,
      plan: ctx!.otherPlan,
      merged: ctx!.otherMerged,
      activeProfileName: report.active,
    });
    expect(result.action).toBe("aborted");

    const stateAfter = await readStateFile(paths);
    expect(stateAfter.state.activeProfile).toBe(stateBefore.state.activeProfile);
    expect(
      await fs.readFile(path.join(paths.claudeDir, "CLAUDE.md"), "utf8"),
    ).toBe(beforeBytes);
  });

  it("S15-extension: persist split-brain — if killed mid-flow, next swap reconciles cleanly", async () => {
    // The persist transactional pair (R22b) writes via pending/prior. If the
    // process dies between persist completion and materialize, the next CLI
    // run reconciles. We simulate this by leaving a `.pending/` dir under
    // the active profile and verifying that a follow-up persist produces a
    // consistent result.
    const paths = buildStatePaths(ctx!.fx.projectRoot);
    await materialize(paths, ctx!.leafPlan, ctx!.leafMerged);
    await fs.writeFile(
      path.join(paths.claudeDir, "CLAUDE.md"),
      "EDITED\n",
    );

    // Inject a stale persist-side `.pending/` to simulate a half-finished
    // prior run.
    const profileDir = path.join(
      ctx!.fx.projectRoot,
      ".claude-profiles",
      "leaf",
    );
    const stalePending = path.join(profileDir, ".pending");
    await fs.mkdir(stalePending, { recursive: true });
    await fs.writeFile(path.join(stalePending, "STALE_GHOST"), "ghost");

    // Now perform the persist gate normally; the orchestration should
    // reconcile (drop stale .pending) before staging fresh content. We don't
    // expect any STALE_GHOST in the final persisted profile.
    const result = await applyGate("persist", {
      paths,
      plan: ctx!.otherPlan,
      merged: ctx!.otherMerged,
      activeProfileName: "leaf",
    });
    expect(result.action).toBe("persisted-and-materialized");

    const persistedDir = path.join(profileDir, ".claude");
    const ghost = path.join(persistedDir, "STALE_GHOST");
    await expect(fs.stat(ghost)).rejects.toMatchObject({ code: "ENOENT" });
    expect(
      await fs.readFile(path.join(persistedDir, "CLAUDE.md"), "utf8"),
    ).toBe("EDITED\n");
  });

  it("non-interactive auto-abort flow: drift + non-interactive + no flag → no change", async () => {
    const paths = buildStatePaths(ctx!.fx.projectRoot);
    await materialize(paths, ctx!.leafPlan, ctx!.leafMerged);
    await fs.writeFile(
      path.join(paths.claudeDir, "CLAUDE.md"),
      "DRIFT\n",
    );

    const report = await detectDrift(paths);
    const decision = decideGate({ report, mode: "non-interactive" });
    expect(decision.kind).toBe("auto");
    expect(decision.choice).toBe("abort");

    const result = await applyGate(decision.choice!, {
      paths,
      plan: ctx!.otherPlan,
      merged: ctx!.otherMerged,
      activeProfileName: report.active,
    });
    expect(result.action).toBe("aborted");
    expect(
      await fs.readFile(path.join(paths.claudeDir, "CLAUDE.md"), "utf8"),
    ).toBe("DRIFT\n");
  });

  it("clean-swap flow: no drift → no-drift-proceed → materialize", async () => {
    const paths = buildStatePaths(ctx!.fx.projectRoot);
    await materialize(paths, ctx!.leafPlan, ctx!.leafMerged);
    // No edits — clean swap.

    const report = await detectDrift(paths);
    expect(report.entries).toEqual([]);
    const decision = decideGate({ report, mode: "interactive" });
    expect(decision.kind).toBe("no-drift");
    expect(decision.choice).toBe("no-drift-proceed");

    const result = await applyGate(decision.choice!, {
      paths,
      plan: ctx!.otherPlan,
      merged: ctx!.otherMerged,
      activeProfileName: report.active,
    });
    expect(result.action).toBe("materialized");
    expect(result.backupSnapshot).toBeNull();
    const r = await readStateFile(paths);
    expect(r.state.activeProfile).toBe("other");
  });
});

File: go/internal/drift/integration_test.go
package drift_test

import (
	"os"
	"path/filepath"
	"testing"

	"github.com/htxryan/c3p/internal/drift"
	"github.com/htxryan/c3p/internal/state"
)

// S3: drift gate — discard → live tree replaced, edits in snapshot.
func TestIntegration_S3_DiscardReplacesAndSnapshots(t *testing.T) {
	t.Parallel()
	paths, _, otherOpts := setupTwoProfiles(t)
	if err := os.WriteFile(filepath.Join(paths.ClaudeDir, "CLAUDE.md"), []byte("DRIFT FROM USER\n"), 0o644); err != nil {
		t.Fatalf("WriteFile: %v", err)
	}

	report, err := drift.DetectDrift(paths)
	if err != nil {
		t.Fatalf("DetectDrift: %v", err)
	}
	if len(report.Entries) != 1 {
		t.Fatalf("entries = %d, want 1", len(report.Entries))
	}
	decision := drift.DecideGate(drift.GateInput{Report: report, Mode: drift.GateModeInteractive})
	if decision.Kind != drift.GateOutcomePrompt {
		t.Errorf("decision kind = %q, want prompt", decision.Kind)
	}
	// Simulate the user picking discard.
	otherOpts.ActiveProfileName = report.Active
	res, err := drift.ApplyGate(drift.GateChoiceDiscard, otherOpts)
	if err != nil {
		t.Fatalf("ApplyGate: %v", err)
	}
	if res.Action != drift.ApplyActionMaterialized {
		t.Errorf("action = %q, want materialized", res.Action)
	}
	if res.BackupSnapshot == "" {
		t.Fatalf("backupSnapshot empty")
	}
	live, err := os.ReadFile(filepath.Join(paths.ClaudeDir, "CLAUDE.md"))
	if err != nil {
		t.Fatalf("ReadFile live: %v", err)
	}
	if string(live) != "OTHER\n" {
		t.Errorf("live = %q, want OTHER\\n", live)
	}
	got, err := os.ReadFile(filepath.Join(res.BackupSnapshot, "CLAUDE.md"))
	if err != nil {
		t.Fatalf("ReadFile snapshot: %v", err)
	}
	if string(got) != "DRIFT FROM USER\n" {
		t.Errorf("snapshot = %q, want DRIFT FROM USER\\n", got)
	}
	snaps, err := state.ListSnapshots(paths)
	if err != nil {
		t.Fatalf("ListSnapshots: %v", err)
	}
	if len(snaps) != 1 {
		t.Errorf("snaps = %d, want 1", len(snaps))
	}
}

// S4: drift gate — persist → live tree saved into active profile, then swap.
func TestIntegration_S4_PersistSavesIntoActiveProfile(t *testing.T) {
	t.Parallel()
	paths, _, otherOpts := setupTwoProfiles(t)
	if err := os.WriteFile(filepath.Join(paths.ClaudeDir, "CLAUDE.md"), []byte("EDITED LEAF\n"), 0o644); err != nil {
		t.Fatalf("WriteFile edit: %v", err)
	}
	if err := os.WriteFile(filepath.Join(paths.ClaudeDir, "scratch.md"), []byte("scratch\n"), 0o644); err != nil {
		t.Fatalf("WriteFile scratch: %v", err)
	}

	report, err := drift.DetectDrift(paths)
	if err != nil {
		t.Fatalf("DetectDrift: %v", err)
	}
	gotPaths := []string{}
	for _, e := range report.Entries {
		gotPaths = append(gotPaths, e.RelPath)
	}
	want := []string{"CLAUDE.md", "scratch.md"}
	if len(gotPaths) != len(want) || gotPaths[0] != want[0] || gotPaths[1] != want[1] {
		t.Errorf("entry paths = %v, want %v", gotPaths, want)
	}

	otherOpts.ActiveProfileName = report.Active
	res, err := drift.ApplyGate(drift.GateChoicePersist, otherOpts)
	if err != nil {
		t.Fatalf("ApplyGate: %v", err)
	}
	if res.Action != drift.ApplyActionPersistedAndMaterialized {
		t.Errorf("action = %q, want persisted-and-materialized", res.Action)
	}
	persistedDir := filepath.Join(paths.ProfilesDir, "leaf", ".claude")
	if got, _ := os.ReadFile(filepath.Join(persistedDir, "CLAUDE.md")); string(got) != "EDITED LEAF\n" {
		t.Errorf("persisted CLAUDE.md = %q", got)
	}
	if got, _ := os.ReadFile(filepath.Join(persistedDir, "scratch.md")); string(got) != "scratch\n" {
		t.Errorf("persisted scratch = %q", got)
	}
	if got, _ := os.ReadFile(filepath.Join(paths.ClaudeDir, "CLAUDE.md")); string(got) != "OTHER\n" {
		t.Errorf("live = %q, want OTHER\\n", got)
	}
	snaps, err := state.ListSnapshots(paths)
	if err != nil {
		t.Fatalf("ListSnapshots: %v", err)
	}
	if len(snaps) != 0 {
		t.Errorf("snaps = %d, want 0 on persist path", len(snaps))
	}
}

// S6: drift gate — abort → no FS change, state unchanged.
func TestIntegration_S6_AbortLeavesStateUnchanged(t *testing.T) {
	t.Parallel()
	paths, _, otherOpts := setupTwoProfiles(t)
	if err := os.WriteFile(filepath.Join(paths.ClaudeDir, "CLAUDE.md"), []byte("DRIFT\n"), 0o644); err != nil {
		t.Fatalf("WriteFile: %v", err)
	}
	stateBefore, err := state.ReadStateFile(paths)
	if err != nil {
		t.Fatalf("ReadStateFile: %v", err)
	}
	beforeBytes, err := os.ReadFile(filepath.Join(paths.ClaudeDir, "CLAUDE.md"))
	if err != nil {
		t.Fatalf("ReadFile: %v", err)
	}

	report, err := drift.DetectDrift(paths)
	if err != nil {
		t.Fatalf("DetectDrift: %v", err)
	}
	otherOpts.ActiveProfileName = report.Active
	res, err := drift.ApplyGate(drift.GateChoiceAbort, otherOpts)
	if err != nil {
		t.Fatalf("ApplyGate: %v", err)
	}
	if res.Action != drift.ApplyActionAborted {
		t.Errorf("action = %q, want aborted", res.Action)
	}
	stateAfter, err := state.ReadStateFile(paths)
	if err != nil {
		t.Fatalf("ReadStateFile: %v", err)
	}
	if (stateAfter.State.ActiveProfile == nil) != (stateBefore.State.ActiveProfile == nil) {
		t.Errorf("active profile presence changed across abort")
	}
	if stateBefore.State.ActiveProfile != nil && *stateAfter.State.ActiveProfile != *stateBefore.State.ActiveProfile {
		t.Errorf("active profile = %q, want %q (abort should not change state)",
			*stateAfter.State.ActiveProfile, *stateBefore.State.ActiveProfile)
	}
	got, err := os.ReadFile(filepath.Join(paths.ClaudeDir, "CLAUDE.md"))
	if err != nil {
		t.Fatalf("ReadFile after: %v", err)
	}
	if string(got) != string(beforeBytes) {
		t.Errorf("file changed across abort: %q -> %q", beforeBytes, got)
	}
}

// S15-extension: persist split-brain — if killed mid-flow, next swap reconciles.
func TestIntegration_S15Ext_PersistReconcilesAfterStalePending(t *testing.T) {
	t.Parallel()
	paths, _, otherOpts := setupTwoProfiles(t)
	if err := os.WriteFile(filepath.Join(paths.ClaudeDir, "CLAUDE.md"), []byte("EDITED\n"), 0o644); err != nil {
		t.Fatalf("WriteFile: %v", err)
	}

	// Inject a stale persist-side .pending/ to simulate a half-finished
	// prior run.
	profileDir := filepath.Join(paths.ProfilesDir, "leaf")
	stalePending := filepath.Join(profileDir, ".pending")
	if err := os.MkdirAll(stalePending, 0o755); err != nil {
		t.Fatalf("MkdirAll stalePending: %v", err)
	}
	if err := os.WriteFile(filepath.Join(stalePending, "STALE_GHOST"), []byte("ghost"), 0o644); err != nil {
		t.Fatalf("WriteFile ghost: %v", err)
	}

	otherOpts.ActiveProfileName = "leaf"
	res, err := drift.ApplyGate(drift.GateChoicePersist, otherOpts)
	if err != nil {
		t.Fatalf("ApplyGate: %v", err)
	}
	if res.Action != drift.ApplyActionPersistedAndMaterialized {
		t.Errorf("action = %q, want persisted-and-materialized", res.Action)
	}
	persistedDir := filepath.Join(profileDir, ".claude")
	ghost := filepath.Join(persistedDir, "STALE_GHOST")
	if _, err := os.Stat(ghost); !os.IsNotExist(err) {
		t.Errorf("STALE_GHOST should not exist in persisted dir: %v", err)
	}
	if got, _ := os.ReadFile(filepath.Join(persistedDir, "CLAUDE.md")); string(got) != "EDITED\n" {
		t.Errorf("persisted CLAUDE.md = %q, want EDITED\\n", got)
	}
}

// Non-interactive auto-abort flow: drift + non-interactive + no flag → no change.
func TestIntegration_NonInteractiveAutoAbortNoChange(t *testing.T) {
	t.Parallel()
	paths, _, otherOpts := setupTwoProfiles(t)
	if err := os.WriteFile(filepath.Join(paths.ClaudeDir, "CLAUDE.md"), []byte("DRIFT\n"), 0o644); err != nil {
		t.Fatalf("WriteFile: %v", err)
	}

	report, err := drift.DetectDrift(paths)
	if err != nil {
		t.Fatalf("DetectDrift: %v", err)
	}
	decision := drift.DecideGate(drift.GateInput{Report: report, Mode: drift.GateModeNonInteractive})
	if decision.Kind != drift.GateOutcomeAuto {
		t.Errorf("decision kind = %q, want auto", decision.Kind)
	}
	if decision.Choice != drift.GateChoiceAbort {
		t.Errorf("decision choice = %q, want abort", decision.Choice)
	}
	otherOpts.ActiveProfileName = report.Active
	res, err := drift.ApplyGate(decision.Choice, otherOpts)
	if err != nil {
		t.Fatalf("ApplyGate: %v", err)
	}
	if res.Action != drift.ApplyActionAborted {
		t.Errorf("action = %q, want aborted", res.Action)
	}
	if got, _ := os.ReadFile(filepath.Join(paths.ClaudeDir, "CLAUDE.md")); string(got) != "DRIFT\n" {
		t.Errorf("live = %q, want DRIFT\\n", got)
	}
}

// Clean-swap flow: no drift → no-drift-proceed → materialize.
func TestIntegration_CleanSwapNoDriftProceed(t *testing.T) {
	t.Parallel()
	paths, _, otherOpts := setupTwoProfiles(t)
	report, err := drift.DetectDrift(paths)
	if err != nil {
		t.Fatalf("DetectDrift: %v", err)
	}
	if len(report.Entries) != 0 {
		t.Errorf("entries = %d, want 0", len(report.Entries))
	}
	decision := drift.DecideGate(drift.GateInput{Report: report, Mode: drift.GateModeInteractive})
	if decision.Kind != drift.GateOutcomeNoDrift {
		t.Errorf("decision kind = %q, want no-drift", decision.Kind)
	}
	otherOpts.ActiveProfileName = report.Active
	res, err := drift.ApplyGate(decision.Choice, otherOpts)
	if err != nil {
		t.Fatalf("ApplyGate: %v", err)
	}
	if res.Action != drift.ApplyActionMaterialized {
		t.Errorf("action = %q, want materialized", res.Action)
	}
	if res.BackupSnapshot != "" {
		t.Errorf("backupSnapshot = %q, want empty", res.BackupSnapshot)
	}
	r, err := state.ReadStateFile(paths)
	if err != nil {
		t.Fatalf("ReadStateFile: %v", err)
	}
	if r.State.ActiveProfile == nil || *r.State.ActiveProfile != "other" {
		t.Errorf("active profile = %v, want other", r.State.ActiveProfile)
	}
}
